src/node/ledger_secrets.h

```diff
@@ -3,118 +3,283 @@
+      // cached in a unique LedgerSecrets instance that can be accessed without
+      // reading the KV. However, it is possible that the ledger secrets are
+      // updated (e.g. rekey tx) concurrently to their access by another tx. To
+      // prevent conflicts, accessing the ledger secrets require access to a tx
+      // object, which must take a dependency on the secrets table.
+      auto v = tx.get_view(secrets_table);
+
+      // Taking a read dependency on the key at self, which would get updated on
+      // rekey
+      if (!self.has_value())
       {
-        return version == other.version && secret == other.secret;
+        throw std::logic_error(
+          "Node id should be set before taking dependency on secrets table");
       }
-    };
-    // List of ledger secrets that are valid from a specific version to the
-    // version of the next entry in the list. The last entry in the list is
-    // valid for all subsequent versions.
-    std::list<VersionedLedgerSecret> secrets_list;
+      v->get(self.value());
+    }
+
+  public:
+    LedgerSecrets(
+      Secrets& secrets_table_, std::optional<NodeId> self_ = std::nullopt) :
+      secrets_table(secrets_table_),
+      self(self_)
+    {}
 
-    LedgerSecrets() = default;
+    LedgerSecrets(
+      Secrets& secrets_table,
+      NodeId self_,
+      VersionedLedgerSecrets&& ledger_secrets_) :
+      secrets_table(secrets_table),
+      self(self_),
+      ledger_secrets(std::move(ledger_secrets_))
+    {}
 
-    bool operator==(const LedgerSecrets& other) const
+    void init(kv::Version initial_version = 1)
     {
-      return secrets_list == other.secrets_list;
+      std::lock_guard<SpinLock> guard(lock);
+
+      ledger_secrets.secrets.emplace(initial_version, make_ledger_secret());
     }
 
-    void init(kv::Version version = 1)
+    void set_node_id(NodeId id)
     {
-      if (secrets_list.size() != 0)
+      if (self.has_value())
       {
-        throw std::logic_error("Ledger secrets have already been initialised!");
+        throw std::logic_error(
+          "Node id has already been set on ledger secrets");
       }
 
-      secrets_list.push_back({version, LedgerSecret()});
+      self = id;
     }
 
-    LedgerSecret get_latest()
+    VersionedLedgerSecret get_latest(kv::Tx& tx)
     {
-      if (secrets_list.size() == 0)
+      std::lock_guard<SpinLock> guard(lock);
+
+      take_dependency_on_secrets(tx);
+
+      if (ledger_secrets.secrets.empty())
       {
         throw std::logic_error(
           "Could not retrieve latest ledger secret: no secret set");
       }
-      return secrets_list.back().secret;
+
+      auto latest_ledger_secret = ledger_secrets.secrets.rbegin();
+      return std::make_pair(
+        latest_ledger_secret->first, latest_ledger_secret->second);
     }
 
-    std::optional<LedgerSecret> get_penultimate()
+    std::optional<LedgerSecret> get_penultimate(kv::Tx& tx)
```

Comment:
I think we should change this function to fetch both latest and penultimate under the same lock. If a secret is added after fetching one and before fetching the other, we encrypt junk.

Reply:
Good point. Done now.